codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestJsonProtocolGenerator.java

```diff
@@ -374,4 +374,762 @@ private String getSerializedMemberName(MemberShape memberShape) {
+                writer.write("defer response.Body.Close()");
+                writer.write("");
+
+                writer.addUseImports(GoDependency.AWS_JSON_PROTOCOL_ALIAS);
+                writer.write("errorType, message, err := jsonprotocol.GetErrorInfo(decoder)");
+                writer.openBlock( "if len(errortype) == 0 {", "}", () -> {
+                        writer.write("errorType = response.Headers.Get($S)","X-Amzn-Errortype" );
+                });
+                writer.write("errorType = jsonprotocol.SanitizeErrorCode(errorType)");
+
+                writer.openBlock("genericError := $P{","}",
+                        genericAPIErrorSymbol, () ->{
+                    writer.write("Code : errorType,");
+                    writer.write("Message : message,");
+                });
+
+                writer.write("return out, metadata, genericError");
+                return;
+            }
+
+            writer.write("buff := make([]byte, 1024)");
+            writer.write("ringBuffer := sdkio.NewRingBuffer(buff)");
+            writer.write("");
+
+            writer.write("var errorBody bytes.Buffer");
+
+            writer.addUseImports(GoDependency.IO);
+            writer.write("_, err := io.Copy(errorBody, response.Body)");
+            writer.openBlock("if err != nil {", "}", () -> {
+                writer.write(String.format("return out, metadata, &aws.DeserializationError{Err: %s}",
+                        "fmt.Errorf(\"failed to copy error response body, %w\", err)"));
+            });
+            writer.write("body := io.TeeReader(response.Body, ringBuffer)");
+            writer.write("defer response.Body.Close()");
+            writer.write("");
+
+            // Retrieve error shape name from response. For REST JSON protocol, the error shape name can be found either
+            // at Header `X-Amzn-Errortype` or a body field with the name `code`, or a body field named `__type`.
+            writer.addUseImports(GoDependency.JSON);
+            writer.write("decoder := json.NewDecoder(body)");
+            writer.write("decoder.UseNumber()");
+            writer.write("");
+
+            writer.addUseImports(GoDependency.AWS_JSON_PROTOCOL_ALIAS);
+            writer.write("errorType, message, err := jsonprotocol.GetErrorInfo(decoder)");
+            writer.openBlock("if err != nil {", "}", () -> {
+                writer.write("var snapshot bytes.Buffer");
+                writer.write("io.Copy(&snapshot, ringBuffer)");
+                writer.openBlock("return out, metadata, &aws.DeserializationError {", "}", () -> {
+                    writer.write("Err: fmt.Errorf(\"failed to decode response error with invalid JSON, %w\", err),");
+                    writer.write("Snapshot: snapshot.Bytes(),");
+                });
+            });
+
+            writer.openBlock("if len(errorType) == 0 {", "}", () -> {
+                writer.write("errorType = response.Headers.Get($S)", "X-Amzn-Errortype");
+            });
```

Comment:
Nit should check the `X-Amzn-Errortype` header first. If the code/type exist there, you don't need to double decode the error body.  The current impl will double decode the body even if the error code can be found in the header.

Reply:
Even if we get the error type from header, we still need to look for message in the json body. So decided to look for message, and error code in json body first, as that seems more optimal.